xla/service/gpu/buffer_comparator.cu.cc

```diff
@@ -108,141 +108,197 @@ __global__ void xla_fp8_e4m3fnuz_comparison(__hip_fp8_storage_t* buffer_a,
-  if (rel_error > rel_error_threshold || isnan(rel_error))
-    atomicAdd(mismatch_count, 1);
+  int mcount = 0;
+  int unroll = 128 / sizeof(*buffer_a);
+  uint64_t idx = (threadIdx.x + blockIdx.x * blockDim.x) * unroll;
+  for (int i = 0; i < unroll; ++i) {
+    if ((idx+i) < buffer_length) {
+      double elem_a = buffer_a[idx+i];
+      double elem_b = buffer_b[idx+i];
+      if (isnan(elem_a) && isnan(elem_b)) continue;
+      if (isinf(elem_a) && isinf(elem_b) && signbit(elem_a) == signbit(elem_b))
+        continue;
+      double rel_error = abs(elem_a - elem_b) / (max(abs(elem_a), abs(elem_b)) + 1);
+      if (rel_error > rel_error_threshold || isnan(rel_error))
+        mcount++;
+    }
+  }
+  if (mcount)
+    atomicAdd(mismatch_count, mcount);
 }
 
 __global__ void xla_bf16_comparison(bfloat16* buffer_a, bfloat16* buffer_b,
                                     float rel_error_threshold,
                                     uint64_t buffer_length,
                                     int* mismatch_count) {
-  int idx = threadIdx.x + blockIdx.x * blockDim.x;
-  if (idx >= buffer_length) return;
-  float elem_a = BF16_TO_F32(buffer_a[idx]);
-  float elem_b = BF16_TO_F32(buffer_b[idx]);
-  elem_a = Canonicalize(elem_a);
-  elem_b = Canonicalize(elem_b);
-  if (isnan(elem_a) && isnan(elem_b)) return;
-
-  float rel_error = abs(elem_a - elem_b) / (max(abs(elem_a), abs(elem_b)) + 1);
-
-  if (rel_error > rel_error_threshold || isnan(rel_error))
-    atomicAdd(mismatch_count, 1);
+  int mcount = 0;
+  int unroll = 128 / sizeof(*buffer_a);
+  uint64_t idx = (threadIdx.x + blockIdx.x * blockDim.x) * unroll;
+  
+  for (int i = 0; i < unroll; ++i) {
+    if ((idx+i) < buffer_length) {
+      float elem_a = BF16_TO_F32(buffer_a[idx+i]);
+      float elem_b = BF16_TO_F32(buffer_b[idx+i]);
+      elem_a = Canonicalize(elem_a);
+      elem_b = Canonicalize(elem_b);
+      if (isnan(elem_a) && isnan(elem_b)) continue;
+
+      float rel_error = abs(elem_a - elem_b) / (max(abs(elem_a), abs(elem_b)) + 1);
+
+      if (rel_error > rel_error_threshold || isnan(rel_error))
+        mcount++;
+    }
+  }
+  if (mcount)
+    atomicAdd(mismatch_count, mcount);
 }
 
 // TODO(b/191520348): The comparison below requires exact equality.
 __global__ void xla_int8_comparison(int8_t* buffer_a, int8_t* buffer_b,
                                     float rel_error_threshold,
                                     uint64_t buffer_length,
                                     int* mismatch_count) {
-  int idx = threadIdx.x + blockIdx.x * blockDim.x;
-  if (idx >= buffer_length) return;
-  float a = buffer_a[idx];
-  float b = buffer_b[idx];
-  float rel_error = abs(a - b) / (max(abs(a), abs(b)) + 1);
-  if (rel_error > rel_error_threshold || isnan(rel_error))
-    atomicAdd(mismatch_count, 1);
+  int mcount = 0;
+  int unroll = 128;
+  int idx = (threadIdx.x + blockIdx.x * blockDim.x) * unroll;
```

Comment:
Can you please change the type of 'idx' and 'unroll' to uint64_t for all kernels ? 
I am just afraid of that the compiler will first calculate the product 'threadIdx.x + blockIdx.x * blockDim.x) * unroll' using int32 arithmetic and then cast it to uint64_t ...